proposals/0000-unlifted-datatypes.rst

```diff
@@ -0,0 +1,482 @@
+declaration introduces a *lifted* type of kind ``TYPE (BoxedRep Lifted)``.
+
+So omission of the zero tag check is only possible if ⊥ is not an inhabitant of
+the type. This is exactly the condition for a type to be *unlifted*! Since we
+are still talking about heap objects, we need unlifted, but boxed types, which
+have kind ``TYPE (BoxedRep Unlifted)``. Indeed, for types of this kind GHC will
+already omit the zero tag check today! So all that is left is to extend GHC in
+a way that
+
+::
+  
+  data IntSet :: TYPE (BoxedRep Unlifted) where
+    Branch :: IntSet -> !Int -> IntSet -> IntSet
+    Leaf   :: IntSet
+
+  member :: Int -> IntSet -> Bool
+  member !k Leaf = False
+  member k (Branch l k' r)
+    | k == k'   = True
+    | k <  k'   = member k l
+    | otherwise = member k r
+
+does the Right Thing, namely disallowing unevaluated boxes to be passed around
+by using call-by-value. This way we can have the guarantee that all unlifted
+boxes are always properly tagged.
+
+Preliminary `benchmarks <https://gitlab.haskell.org/ghc/ghc/merge_requests/2218#note_239994>`_
+suggest that the unlifted variant is 5-10% faster, just due to omission of the
+zero tag check.
+
+Proposed Change Specification
+-----------------------------
+
+Henceforth, data type declaration refers to both data type and data family instance declarations.
+
+Static semantics
+~~~~~~~~~~~~~~~~
+
+Kind signatures for regular data type declarations must have a return kind of
+``TYPE (BoxedRep Lifted)``. Activating ``-XUnliftedDatatypes`` will lift this
+restriction to allow any return kind that unifies with ``TYPE (BoxedRep _)``. A
+data type explitly *may* have levity polymorphic kind
+``TYPE (BoxedRep l)`` for some type variable ``l``.
+
+Note that Haskell98-style data declarations can use standalone kind signatures
+to specify the return kind. If the user provides no kind signature, the default
+remains that the data type has lifted return kind (``TYPE (BoxedRep Lifted)``).
+Specifying a kind signature is the only way to declare an unlifted or levity-polymorphic data type.
+
+The same applies to data family instances, in which case the data family
+application's result kind must reduce to ``TYPE (BoxedRep _)``. See
+`the section on data families in the UnliftedNewtypes proposal <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0098-unlifted-newtypes.rst>`_
+and ``Note [Implementation of UnliftedNewtypes]`` for details involving
+type-checking the parent data family.
+
+As usual for types of unlifted kind, values of unlifted data types may not be
+declared at the top-level. The top-level restriction is not fundamental (see
+`#17521 <https://gitlab.haskell.org/ghc/ghc/issues/17521>`_), but best
+discussed in a separate proposal. Note that we don't consider nullary
+constructors to be top-level bindings in that regard, so nullary unlifted
+constructors are allowed. There *is* an interesting interaction of this
```

Comment:
Example
```
zs :: SList a
zs = SNil
```
I have no idea why you want to allow this.  It seems... odd.

Reply:
That's exactly what is not allowed. I wrote

>  we don't consider nullary constructors to be top-level bindings in that regard, so nullary unlifted constructors are allowed

to mean that nullary constructors aren't affected by the top-level restriction, so defining a nullary unlifted datacon is just fine. I'll change that to

> we don't consider nullary constructors to be top-level bindings in that regard, so *defining* nullary unlifted constructors *is* allowed

Happy to use different words here if that's still confusing, but I don't know which.

Or maybe just get rid of the section beginning at "Note that we don't"?

Reply:
> we don't consider nullary constructors to be top-level bindings in that regard, so defining nullary unlifted constructors is allowed

Oh I totally misunderstood.   You mean that
```
data T :: TYPE (BoxedRep Unlifted) where
  T1 :: T
```
is allowed.  Yes, indeed.  It never occurred to me that you might consider that that to be a "top level binding".   I thought you meant that `t1 = T1` is allowed!  But you don't.   

I think you can just say "As usual for types of unlifted kind, values of unlifted data types may not be declared at the top-level."; then give an example of what is and is not allowed (I offered a suggestion); then your following sentence "The top-level restriction is not fundamental (see #17521), but best discussed in a separate proposal."  Then stop

Reply:
Done